File: 单调栈/Next Greater Element II.java
class Solution {
    public int[] nextGreaterElements(int[] nums) {
        int N = nums.length;
        int[] res = new int[N];
        Arrays.fill(res, -1);
        Stack<Integer> stack = new Stack<>();
        for (int i = 0; i < N * 2; i++) {
            // 这里不可以加上等于号，原因是我的题目要求是一定是大于，不可能是等于
            while (!stack.isEmpty() && nums[i % N] > nums[stack.peek() % N]) {
                res[stack.pop() % N] = nums[i % N];
            }
            stack.push(i % N);
        }
        return res;
    }
}

File: 单调栈/next-greater-element-ii.cpp
class Solution {
public:
// 遇到这种循环遍历的题目一般是拼接数组, 这里采用模拟拼接的方式
    vector<int> nextGreaterElements(vector<int>& nums) {
        int N = nums.size();
        // 先把值初始化成-1
        vector<int> res(N, -1);
        // 单调栈
        stack<int> st;
        
        for (int i = 0; i < 2*N; i++) {
            while (!st.empty() && nums[st.top()] < nums[i%N]) {
                // 得到当前的值
                int x = st.top();
                res[x] = nums[i%N];
                st.pop();
            }
            st.push(i % N); 
        }
        return res;
    }
};
